src/wallet/wallet2.h

```diff
@@ -1927,6 +1942,34 @@ namespace boost
         return;
       a & x.multisig_sigs;
     }
+
+    template <class Archive>
+    inline typename std::enable_if<!Archive::is_loading::value, void>::type serialize(Archive &a, boost::container::flat_map<crypto::public_key, cryptonote::subaddress_index> &x, const boost::serialization::version_type ver)
+    {
+      uint32_t sz = x.size();
+      a & sz;
```

Comment:
Possibly. Why ?

Reply:
The code for version change may not be necessary if both data structures have the same binary input/output. Its a matter of loading the same data into a different C++ data structure.

Reply:
OK. Looking at both, I think it's binary compatible indeed.

Reply:
Looks like it's not. The obvious difference was size_t/uint32_t, but after changing that it's still not. If you fancy working out why feel free, but I'm not wasting time on that.

Reply:
The `uint32_t` was a good catch, it needs to be `size_t` here for correctness.

> but I'm not spending more time on that since the gain is just saving a few lines.

Unfortunately, I should've read Boost code before assuming its implementation. Boost is storing the bucket count too, despite it not being strictly necessary, and there is no such concept with `flat_map`.

Reply:
That should not matter, since monero code implements that AFAICT.

Reply:
I don't understand this response given the context of my last comment. This appears to be `std::size_t` now or - ?

Reply:
Yes, I changed it to size_t, but this was not enough to make it binary compatible. My comment was "inclusion of bucket count should not matter since monero uses its own code to serialize AFAIK".

Reply:
I was not concise in that comment.

The `boost::serialization` code for `std::unordered_map` writes out a bucket count to the stream. That is one of the reasons for the binary incompatibility. No such concept (bucket count) exists with `flat_map`, so making it binary compatible is a stretch.

Reply:
So you're saying it's not using the code in src/common/unordered_containers_boost_serialization.h, lines 41 to 67, but some other code ? Because the code here does not save a bucket count AFAICT.

Reply:
I wasn't aware of this specialization, I thought the Boost default was being used.

When you tested compatibility, did you do a direct comparison of the binary outputs? The two will differ - `unordered_map` has undefined order whereas `flat_map` is guaranteed to be in-order. I did a quick test (directly, not via the entire wallet2 specification), and storing `unordered_map` can be loaded by `flat_map`. There is one catch - if you use my `x.insert(x.end(), e)` technique, the first load will be slower since the elements are not in-order until the next store.

Reply:
No, I loaded a cache saved with unordered_map with the flat_map code.

Reply:
It might be saving the type first. That'd explain both why it loads for you and not for me.